File: packages/react-native-worklets/plugin/src/bundleMode.ts
import type { NodePath } from '@babel/core';
import { booleanLiteral, type ExpressionStatement } from '@babel/types';
import { join } from 'path';

import type { WorkletsPluginPass } from './types';

const WORKLETS_PACKAGE = 'react-native-worklets';
const WORKLETS_SRC_DIR = join(WORKLETS_PACKAGE, 'src');
const WORKLETS_LIB_DIR = join(WORKLETS_PACKAGE, 'lib', 'module');

const togglePaths = [
  join(WORKLETS_SRC_DIR, 'index.ts'),
  join(WORKLETS_SRC_DIR, 'debug', 'bundleMode.native.ts'),
  join(WORKLETS_LIB_DIR, 'index.js'),
  join(WORKLETS_LIB_DIR, 'debug', 'bundleMode.native.js'),
];

/**
 * This function replaces the `false` value in
 *
 * `globalThis._WORKLETS_BUNDLE_MODE_ENABLED = false;`
 *
 * With `true` in each of the Worklets' toggle-target files when the
 * `bundleMode` option is enabled in the Babel plugin.
 *
 * This way Bundle Mode is not accidentally set up in eager import environments.
 */
export function toggleBundleMode(
  path: NodePath<ExpressionStatement>,
  state: WorkletsPluginPass
) {
  if (
    !state.opts.bundleMode ||
    !togglePaths.some((togglePath) => state.filename?.endsWith(togglePath))
  ) {
    return;
  }

  const expressionPath = path.get('expression');

  if (!expressionPath.isAssignmentExpression()) {
    return;
  }

  const left = expressionPath.get('left');

  if (!left.isMemberExpression()) {
    return;
  }

  const object = left.get('object');
  const property = left.get('property');

  if (
    !object.isIdentifier() ||
    object.node.name !== 'globalThis' ||
    !property.isIdentifier() ||
    property.node.name !== '_WORKLETS_BUNDLE_MODE_ENABLED'
  ) {
    return;
  }

  const right = expressionPath.get('right');

  right.replaceWith(booleanLiteral(true));
}

File: packages/react-native-worklets/plugin-oxc/src/transformer/bundle_mode.rs
use oxc_ast::AstBuilder;
use oxc_ast::ast::{AssignmentTarget, Expression, ExpressionStatement};
use oxc_span::SPAN;

use crate::state::State;

const TOGGLE_PATHS: &[&str] = &[
    "react-native-worklets/src/index.ts",
    "react-native-worklets/src/debug/bundleMode.native.ts",
    "react-native-worklets/lib/module/index.js",
    "react-native-worklets/lib/module/debug/bundleMode.native.js",
];

pub fn toggle_bundle_mode<'a>(
    node: &mut ExpressionStatement<'a>,
    _state: &State,
    filename: &str,
    builder: AstBuilder<'a>,
) {
    if !TOGGLE_PATHS.iter().any(|path| filename.ends_with(path)) {
        return;
    }

    let Expression::AssignmentExpression(assign) = &mut node.expression else {
        return;
    };

    let AssignmentTarget::StaticMemberExpression(member) = &assign.left else {
        return;
    };

    let Expression::Identifier(object) = &member.object else {
        return;
    };
    if object.name.as_str() != "globalThis" {
        return;
    }
    if member.property.name.as_str() != "_WORKLETS_BUNDLE_MODE_ENABLED" {
        return;
    }

    assign.right = builder.expression_boolean_literal(SPAN, true);
}
